hkmc2/shared/src/main/scala/hkmc2/codegen/StackSafeTransform.scala

```diff
@@ -0,0 +1,189 @@
+package hkmc2
+
+import mlscript.utils.*, shorthands.*
+import hkmc2.codegen.*
+import hkmc2.semantics.Elaborator.State
+import hkmc2.semantics.*
+import hkmc2.syntax.Tree
+import hkmc2.utils.*
+
+class StackSafeTransform(depthLimit: Int)(using State):
+  private val STACK_DEPTH_IDENT: Tree.Ident = Tree.Ident("__stackDepth")
+  private val STACK_OFFSET_IDENT: Tree.Ident = Tree.Ident("__stackOffset")
+  private val STACK_HANDLER_IDENT: Tree.Ident = Tree.Ident("__stackHandler")
+
+  private val stackDelayClsPath: Path = State.globalThisSymbol.asPath.selN(Tree.Ident("Predef")).selN(Tree.Ident("__StackDelay")).selN(Tree.Ident("class"))
+  private val stackDepthPath: Path = State.globalThisSymbol.asPath.selN(Tree.Ident("Predef")).selN(STACK_DEPTH_IDENT)
+  private val stackOffsetPath: Path = State.globalThisSymbol.asPath.selN(Tree.Ident("Predef")).selN(STACK_OFFSET_IDENT)
+  private val stackHandlerPath: Path = State.globalThisSymbol.asPath.selN(Tree.Ident("Predef")).selN(STACK_HANDLER_IDENT)
+  private val predefPath: Path = State.globalThisSymbol.asPath.selN(Tree.Ident("Predef"))
+
+  private def intLit(n: BigInt) = Value.Lit(Tree.IntLit(n))
+  
+  private def op(op: String, a: Path, b: Path) =
+    Call(State.builtinOpsMap(op).asPath, List(a.asArg, b.asArg))(true)
+
+  // TODO: this code is copied from HandlerLowering and is quite useful. Maybe refactor it into a utils file
```

Comment:
Yes. Let's not add duplicated code, which will need twice the maintenance burden and may diverge.